### PR TITLE
Tool/crosscheckfingerprints

### LINKED ISSUE #1502: new module: picard/CrossCheckFingerprints
I think it would be good to have a module for [Crosscheckfingerprints](https://gatk.broadinstitute.org/hc/en-us/articles/360037270871-CrosscheckFingerprints-Picard-)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by matthdsm**
## PR checklist

Closes #1502

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/picard/crosscheckfingerprints/main.nf [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    path "versions.yml"                                 ,emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    def input1_string = input1.join(" --INPUT ")
```
Comment (sateeshperi) [inline modules/picard/crosscheckfingerprints/main.nf:26]: @matthdsm  i'm curious as to why `--INPUT` and `--SECOND_INPUT` are being joined to the inputs as to declaring them as params on L41 (actually wouldn't it be duplicating here!) and L42 ?

Comment (matthdsm) [reply modules/picard/crosscheckfingerprints/main.nf:26]: Hi,
`input1` is a mandatory field for the tool,  so it makes sense (to me) to code that in the script part.
`input2` however is optional, so we need to need to make sure it exists. If it has one value the `join` won't do anything, so I add `--SECOND_INPUT` before it.

I'm still new to nextflow and groovy, so no idea if there's a better way to do this. All suggestions are welcome!

File: modules/picard/crosscheckfingerprints/main.nf [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        'quay.io/biocontainers/picard:2.26.10--hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(input1)
+    path(input2)
+    file haplotype_map
+
+    output:
+    tuple val(meta), path("*.crosscheck_metrics.txt")   ,emit: crosscheck_metrics
+    path "versions.yml"                                 ,emit: versions
```
Comment (JoseEspinosa) [inline modules/picard/crosscheckfingerprints/main.nf:17]:
```suggestion
    tuple val(meta), path("*.crosscheck_metrics.txt"), emit: crosscheck_metrics
    path "versions.yml"                              , emit: versions
```

File: modules/picard/crosscheckfingerprints/main.nf [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    conda (params.enable_conda ? "bioconda::picard=2.26.10" : null)
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/picard:2.26.10--hdfd78af_0' :
+        'quay.io/biocontainers/picard:2.26.10--hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(input1)
+    path(input2)
+    file haplotype_map
```
Comment (JoseEspinosa) [inline modules/picard/crosscheckfingerprints/main.nf:13]:
```suggestion
    path input2
    path haplotype_map
```

File: tests/modules/picard/crosscheckfingerprints/main.nf [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+process STUB_HAPLOTYPE_MAP {
+
+    output:
+    path "haplotype_map.txt" ,emit:haplotype_map
+
+    stub:
+    """
+    echo haplotype_map > haplotype_map.txt
+    """
+}
```
Comment (JoseEspinosa) [inline tests/modules/picard/crosscheckfingerprints/main.nf:16]: For the stub run you can just declare a channel empty for the `haplotype_map` input I think, as shown in the next  suggestion
```suggestion
```

File: tests/modules/picard/crosscheckfingerprints/main.nf [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+workflow test_picard_crosscheckfingerprints {
+
+    input = [
+        [ id:'test', single_end:false ], // meta map
+        [file(params.test_data['sarscov2']['illumina']['test_paired_end_bam'], checkIfExists: true), file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExists: true)],
+    ]
+    STUB_HAPLOTYPE_MAP ()
+    PICARD_CROSSCHECKFINGERPRINTS ( input,[],STUB_HAPLOTYPE_MAP.out.haplotype_map )
```
Comment (JoseEspinosa) [inline tests/modules/picard/crosscheckfingerprints/main.nf:25]:
```suggestion
    PICARD_CROSSCHECKFINGERPRINTS ( input, [], [] )
```

**Review by JoseEspinosa — CHANGES_REQUESTED** [review]
Nice @matthdsm !
You can keep the `stub-run` on the test but the idea is that test actually check the real function of the module. Wouldn't it be possible to create a small `haplotype_map` file to run the tests. In [this](https://nf-co.re/developers/modules#test-data) section of the docs it is explained how to add test files to the test-datasets.

File: tests/modules/picard/crosscheckfingerprints/main.nf (hunk 5 above)
Comment (matthdsm) [reply tests/modules/picard/crosscheckfingerprints/main.nf:25]: Can't do this, the haplotype map is a mandatory argument for this tool

File: tests/modules/picard/crosscheckfingerprints/main.nf (hunk 4 above)
Comment (matthdsm) [reply tests/modules/picard/crosscheckfingerprints/main.nf:16]: Can't do this, the haplotype map is a mandatory argument for this tool

File: modules/picard/crosscheckfingerprints/main.nf (hunk 3 above)
Comment (matthdsm) [reply modules/picard/crosscheckfingerprints/main.nf]: Whoops, that was messy

**Comment by matthdsm** [comment]
> Nice @matthdsm !
> You can keep the `stub-run` on the test but the idea is that test actually check the real function of the module. Wouldn't it be possible to create a small `haplotype_map` file to run the tests. In [this](https://nf-co.re/developers/modules#test-data) section of the docs it is explained how to add test files to the test-datasets.

Hi,
I'm not sure. For this to work we'd need some snp's in the bam, the prevalence in the population of that snp and a similar sample. I'll check if I can mock something up

File: tests/modules/picard/crosscheckfingerprints/main.nf (hunk 5 above)
Comment (JoseEspinosa) [reply tests/modules/picard/crosscheckfingerprints/main.nf]: I see, passing an empty channel worked with the `stub-run` but for testing the module you need a real file I see

**Comment by matthdsm** [comment]
Needs input data:
https://github.com/nf-core/test-datasets/pull/499

**Comment by matthdsm** [comment]
@JoseEspinosa, care to re-review? The tool actually runs now 😄

File: modules/picard/crosscheckfingerprints/main.nf [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        $args \\
+        --NUM_THREADS ${task.cpus} \\
+        --INPUT $input1_string \\
+        $input2_string \\
+        --HAPLOTYPE_MAP ${haplotype_map} \\
+        --OUTPUT ${prefix}.crosscheck_metrics.txt
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        picard: \$(echo \$(picard CrosscheckFingerprints --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d:)
```
Comment (sateeshperi) [inline modules/picard/crosscheckfingerprints/main.nf:48]:
```suggestion
        picard: \$( picard CrosscheckFingerprints --version 2>&1 | grep -o 'Version:.*' | cut -f2- -d: )
```

File: modules/picard/crosscheckfingerprints/main.nf [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    """
+
+    stub:
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    echo ${prefix}.crosscheck_metrics > ${prefix}.crosscheck_metrics.txt
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        picard: \$(echo \$(picard CrosscheckFingerprints --version 2>&1) | grep -o 'Version:.*' | cut -f2- -d:)
```
Comment (sateeshperi) [inline modules/picard/crosscheckfingerprints/main.nf:59]:
```suggestion
        picard: \$( picard CrosscheckFingerprints --version 2>&1 | grep -o 'Version:.*' | cut -f2- -d: )
```

**Review by sateeshperi — APPROVED** [review]
I feel the `echo` is unnecessary.....rest LGTM

File: modules/picard/crosscheckfingerprints/main.nf (hunk 7 above)
Comment (matthdsm) [reply modules/picard/crosscheckfingerprints/main.nf:59]: You're right!

**Review by JoseEspinosa — APPROVED** [review]
Thanks @matthdsm! 🚀
You could also have left the workflow for testing the `stub-run` mode  but up to you to add it again

**Comment by matthdsm** [comment]
No particular fan of the stub mode when not needed, we already know  `echo` works, so no needs to test it into oblivion 😅 
Do you feel it's good to merge?

**Pull request merged**